utils/providers.py

```diff
@@ -1,388 +1,437 @@
-        added_providers.extend(_setup_providers(pclass, **setup_kwargs))
-
-    if validate:
-        for provider in added_providers:
-            provider.validate()
-
-    perflog.stop('utils.providers.setup_providers')
-
-    return added_providers
+    for provider_key, provider_data in providers_data.items():
+        if provider_data.get("name") == provider_name:
+            return get_crud(provider_key)
+    raise NameError("Could not find provider {}".format(provider_name))
 
 
-def _setup_providers(prov_class, validate, check_existing):
-    """Helper to set up all cloud, infra or container providers, and then validate them
+def get_mgmt(provider_key, providers=None, credentials=None):
+    """ Provides a ``mgmtsystem`` object, based on the request.
 
     Args:
-        prov_class: Provider class - 'cloud, 'infra', 'container' or 'middleware' (a string)
-        validate: see description in :py:func:`setup_provider`
-        check_existing: see description in :py:func:`setup_provider`
-
-    Returns:
-        A list of provider objects that have been created.
-
+        provider_key: The name of a provider, as supplied in the yaml configuration files.
+            You can also use the dictionary if you want to pass the provider data directly.
+        providers: A set of data in the same format as the ``management_systems`` section in the
+            configuration yamls. If ``None`` then the configuration is loaded from the default
+            locations. Expects a dict.
+        credentials: A set of credentials in the same format as the ``credentials`` yamls files.
+            If ``None`` then credentials are loaded from the default locations. Expects a dict.
+    Return: A provider instance of the appropriate ``mgmtsystem.MgmtSystemAPIBase``
+        subclass
     """
-
-    # Check for existing providers all at once, to prevent reloading
-    # the providers page for every provider in cfme_data
-    if not list_providers(BaseProvider.type_mapping[prov_class].provider_types.keys()):
-        return []
-    if check_existing:
-        navigate = "{}_providers".format(
-            BaseProvider.type_mapping[prov_class].page_name)
-        sel.force_navigate(navigate)
-        add_providers = []
-        for provider_key in list_providers(
-                BaseProvider.type_mapping[prov_class].provider_types.keys()):
-            provider_name = conf.cfme_data.get('management_systems', {})[provider_key]['name']
-            quad_name = BaseProvider.type_mapping[prov_class].quad_name
-            quad = Quadicon(provider_name, quad_name)
-            for page in paginator.pages():
-                if sel.is_displayed(quad):
-                    logger.debug('Provider %s exists, skipping', provider_key)
-                    break
-            else:
-                add_providers.append(provider_key)
+    if providers is None:
+        providers = providers_data
+    # provider_key can also be provider_data for some reason
```

Comment:
This was because I was making an artificial provider from one of the vsphere's host, therefore I needed a way how to do that.